sphinx/builders/html.py

```diff
@@ -404,15 +404,22 @@ def get_doc_context(self, docname, body, metatags):
         # title rendered as HTML
         title = self.env.longtitles.get(docname)
         title = title and self.render_partial(title)['title'] or ''
+
+        # Suffix for the document
+        source_suffix = path.splitext(self.env.doc2path(docname))[1]
+
         # the name for the copied source
-        sourcename = self.config.html_copy_source and docname + '.txt' or ''
+        if self.config.html_copy_source:
+            if self.config.html_sourcelink_keep_suffix:
+                sourcename = docname + source_suffix
+            else:
+                sourcename = docname + '.txt'
```

Comment:
While we're here, we might also want to change the default behaviour:  With multiple input file types now being allowed, it might make sense to generate `myinput.rst.txt` and `myotherinput.md.txt` now, instead, so that users can get a clue what the original file *type* was.